token-repository/src/main/scala/io/renku/tokenrepository/repository/association/TokenAssociator.scala

```diff
@@ -53,18 +58,49 @@ private class TokenAssociatorImpl[F[_]: MonadThrow](
 
   private def encryptAndPersist(projectId: Id, projectPath: Path, token: AccessToken) = for {
     encryptedToken <- encrypt(token)
-    _              <- persistAssociation(projectId, projectPath, encryptedToken)
+    _              <- persistOrRetry(projectId, projectPath, encryptedToken)
   } yield ()
+
+  private def persistOrRetry(projectId:       Id,
+                             projectPath:     Path,
+                             encryptedToken:  AccessTokenCrypto.EncryptedAccessToken,
+                             numberOfRetries: Int = 0
+  ): F[Unit] = for {
+    _ <- persistAssociation(projectId, projectPath, encryptedToken)
+    _ <- verifyTokenIntegrity(projectPath) recoverWith retry(projectId, projectPath, encryptedToken, numberOfRetries)
+  } yield ()
+
+  private def verifyTokenIntegrity(projectPath: Path) =
+    tokenFinder.findToken(projectPath).value.flatMap {
+      case Some(savedToken) => decrypt(savedToken).void
+      case _ =>
+        new Exception(show"Token associator - saved encrypted token cannot be found for project: $projectPath")
```

Comment:
Should we maybe raise the exception only when we run out of attempts?

Reply:
Yes good idea!